aten/src/ATen/gen.py

```diff
@@ -390,9 +390,6 @@ def legacy_iterate_types():
             for scalar_type in (scalar_types + quantized_scalar_types):
                 if density == 'Mkldnn' and (backend != 'CPU' or scalar_type[0] != 'Float'):
                     continue
-                if density == 'Sparse' and scalar_type[0] == 'Half':
```

Comment:
this means we have CPU + Sparse + Half tensors, right?  I'd expect some minimal tests of:
1) constructing cpu + sparse + half, cuda + sparse + half tensors (in `test_sparse.py`).
2) converting to/from them (across dtype, across device boundaries), copying (also in `test_sparse.py`).

Reply:
Yes this allows for construction of sparse half cpu tensors, though backward doesn't work (yet).

It seems like this should be in test_nn not test_sparse, as there isn't any use of embeddings in test_sparse currently? The test_nn tests added should cover some of those construction conversion cases (in test_nn  we move cpu->cuda and float->half already just to perform the backward tests) and easy to expand existing construction/copy tests in test_nn to include the half type.

Reply:
The main reason we end up needing cpu+sparse+half is just that we can't create a half sparse embedding on cuda directly (Embedding doesn't take a dtype parameter). Instead we call `Embedding(x, y, sparse=True).half().cuda()`.

Reply:
added/updated tests in test_sparse.py and fixed issue with construction of torch.sparse.HalfTensor